### PR TITLE
Align parser and ast with new grammar

### PR DESCRIPTION
**Pull request by jens270996**
Extended AST with new grammar supporting procedures.

Then moved on to "fixing" all the errors that occurred as a result of this change.
This was done by either
 - Fixing error, in cases where the fix was trivial
 - Commenting out function/case and substituting with undefined in non-trivial cases.

Following PRs should be alot smaller, since now it is possible to start implementing the extended version of ARL incrementally.
e.g. implementing the parser can be done without having to touch the interpreter or PE modules.

### CODE REVIEW

File: src/PE/Preprocessing/Division.hs [hunk 1]
```diff
@@ -46,6 +46,7 @@ makeDiv store decl =
 
 startingDiv :: (Ord a) => NormProgram a -> PWDivision a
                               -> Division
-startingDiv (_, p) pwd =
-  let n = nname $ getNEntryBlock p
-  in fst $ get n pwd
+startingDiv p pwd =
+  undefined -- TODO: fix
```
Comment (jens270996) [inline src/PE/Preprocessing/Division.hs:50]:
```suggestion
   undefined --TODO: fix when adding PE support for processes
```

File: src/PE/Preprocessing/Impl/Annotate.hs [hunk 2]
```diff
@@ -45,19 +48,21 @@ annotateStep _ _ Skip = Skip' BTStatic
 
 -- Annotate a come-from
 annotateFrom :: Division -> ComeFrom a () -> ComeFrom' a
-annotateFrom _ (Entry ()) = Entry'
-annotateFrom _ (From (l, ())) = From' l
-annotateFrom d (Fi e (l1, ()) (l2, ())) =
-  let (e', btType) = annotateExp d e
-  in Fi' btType e' l1 l2
+annotateFrom = undefined -- TODO: fix
```
Comment (jens270996) [inline src/PE/Preprocessing/Impl/Annotate.hs:51]:
```suggestion
annotateFrom  = undefined --TODO: fix when adding PE support for processes
```

File: src/PE/Preprocessing/Impl/Uniform.hs [hunk 3]
```diff
@@ -16,11 +16,12 @@ type ST a = S.State Division a
 
 -- Create a congruent uniform division for a program
 congruentUniformDiv :: Ord a => NormProgram a -> Division -> PWDivision a
-congruentUniformDiv (decl, p) d =
-  let congruentDiv = makeCongruent (decl, p) d
-      ls = map nname p
-      pairs = map (\l -> (l, (congruentDiv, congruentDiv))) ls
-  in fromList pairs
+congruentUniformDiv p d =
+  undefined -- TODO: fix
```
Comment (jens270996) [inline src/PE/Preprocessing/Impl/Uniform.hs:20]:
```suggestion
   undefined --TODO: fix when adding PE support for processes
```

File: src/PE/Preprocessing/Impl/Uniform.hs [hunk 4]
```diff
@@ -33,7 +34,8 @@ makeCongruent p = fixed (S.execState $ checkProg p)
 -- The Division is preserved in the state-monad for fewer arguments passed
 -- and the division will grow increasingly dynamic throughout the iteration
 checkProg :: NormProgram a -> ST ()
-checkProg (decl, p) = mapM_ (checkBlock decl) p
+checkProg p = undefined -- TODO: fix
```
Comment (jens270996) [inline src/PE/Preprocessing/Impl/Uniform.hs:37]:
```suggestion
checkProg p  = undefined --TODO: fix when adding PE support for processes
```

File: src/Parsing/Impl/Program.hs [hunk 5]
```diff
@@ -16,15 +16,10 @@ parseProg = parseStr pProg
...
--- parse a variable declaration
-pDecl :: Parser VariableDecl
-pDecl =
-  VariableDecl <$> pNames <*> (symbol "->" *> pNames)
-               <*> option [] (word "with" *> pNames)
-  where pNames = symbol "(" *> many pName <* symbol ")"
+pProg = many1 pProcess
 
+pProcess :: Parser (Process Label ())
+pProcess = undefined
```
Comment (jens270996) [inline src/Parsing/Impl/Program.hs:22]:
```suggestion
pProcess  = undefined --TODO: fix when adding parsing for processes
```

File: src/RL/AST.hs [hunk 6]
```diff
@@ -106,10 +122,10 @@ mapProgram f g = map changeBlock
       , jump = appJump $ jump b
       }
     appName (l, s) = (f l s, g s)
-    appFrom (Entry s) = Entry (g s)
+    appFrom (Entry p s) = Entry p (g s)
```
Comment (jens270996) [inline src/RL/AST.hs:125]:
```suggestion
    appFrom (Entry p s) = Entry p (g s)  --TODO: is correct?

File: src/RL/AST.hs [hunk 7]
```diff
@@ -106,10 +122,10 @@ mapProgram f g = map changeBlock
...
     appFrom (From (l, s)) = From (f l s, g s)
     appFrom (Fi e (l1, s1) (l2, s2)) = Fi e (f l1 s1, g s1) (f l2 s2, g s2)
-    appJump (Exit s) = Exit (g s)
+    appJump (Exit p s) = Exit p (g s)
```
Comment (jens270996) [inline src/RL/AST.hs:128]:
```suggestion
    appJump (Exit p s) = Exit p (g s) -- TODO: is correct?
```

File: src/RL/AST.hs [hunk 8]
```diff
@@ -124,12 +140,12 @@ mapBlock f b = b
   }
 
 mapFrom :: ((a, b) -> (c, b)) -> ComeFrom a b -> ComeFrom c b
-mapFrom _ (Entry s) = Entry s
+mapFrom _ (Entry p s) = Entry p s
```
Comment (jens270996) [inline src/RL/AST.hs:143]:
```suggestion
mapFrom _ (Entry p s) = Entry p s --TODO: is correct?
```

File: src/RL/AST.hs [hunk 9]
```diff
@@ -124,12 +140,12 @@ mapBlock f b = b
...
 mapFrom f (From l) = From (f l)
 mapFrom f (Fi e l1 l2) = Fi e (f l1) (f l2)
 
 mapJump :: ((a, b) -> (c, b)) -> Jump a b -> Jump c b
-mapJump _ (Exit s) = Exit s
+mapJump _ (Exit p s) = Exit p s
```
Comment (jens270996) [inline src/RL/AST.hs:148]:
```suggestion
mapJump _ (Exit p s) = Exit p s --TODO: is correct?
```

File: src/RL/AST.hs [hunk 10]
```diff
@@ -2,14 +2,22 @@ module RL.AST where
...
 
+--TODO: remove when implementing support for PE with processes
 data VariableDecl = VariableDecl
   { input  :: [Name]
   , output :: [Name]
   , temp   :: [Name]
   } deriving (Eq, Show, Read)
 
+
+data Process label store = Process
```
Comment (Yakokse) [inline src/RL/AST.hs:15]: Process -> Procedure (og alle relaterede steder)
Føler det er en god ide lige at få det rigtige navn på plads før der bliver bygget alt for meget.

File: src/RL/AST.hs [hunk 11]
```diff
@@ -106,10 +122,10 @@ mapProgram f g = map changeBlock
       , jump = appJump $ jump b
       }
     appName (l, s) = (f l s, g s)
-    appFrom (Entry s) = Entry (g s)
+    appFrom (Entry p s) = Entry p (g s)  --TODO: is correct?
```
Comment (Yakokse) [inline src/RL/AST.hs:125]: Ser rigtigt ud! (samme andre steder)

File: src/Utils/PrettyPrint.hs [hunk 12]
```diff
@@ -70,15 +73,15 @@ prettyFrom f (Fi e (l1, ()) (l2, ())) =
   [ "fi " ++ prettyExpr e
   , "\tfrom " ++ f l1
   , "\telse " ++ f l2]
-prettyFrom _ (Entry ()) = ["entry"]
+prettyFrom _ (Entry p ()) = ["entry( " ++ prettyPat p ++ ")"]
```
Comment (Yakokse) [inline src/Utils/PrettyPrint.hs:76]: Nu er vi ikke helt færdig med grammatikken, men ideelt for mig vil det være:
`"entry " ++ prettyPat p`
Samme med exit

**Review by Yakokse — CHANGES_REQUESTED** [review]
Lige et par ting som nævnt på Discord og her.

File: src/Inversion/Impl/Inverter.hs [hunk 13]
```diff
@@ -19,13 +22,13 @@ invertBlock Block {name = l, from = f, body = b, jump = j} = Block
 
 -- invert a come-from
 invertFrom :: ComeFrom a b -> Jump a b
-invertFrom (Entry s)          = Exit s
+invertFrom (Entry p s)          = Exit p s -- TODO: correct?
```
Comment (Yakokse) [inline src/Inversion/Impl/Inverter.hs:25]: Correct👍

File: src/Inversion/Impl/Inverter.hs [hunk 14]
```diff
@@ -19,13 +22,13 @@ invertBlock Block {name = l, from = f, body = b, jump = j} = Block
...
 invertFrom (From l)         = Goto l
 invertFrom (Fi e l1 l2) = If e l1 l2
 
 -- invert a jump
 invertJump :: Jump a b -> ComeFrom a b
-invertJump (Exit s)           = Entry s
+invertJump (Exit p s)           = Entry p s --TODO: correct?
```
Comment (Yakokse) [inline src/Inversion/Impl/Inverter.hs:31]: Correct👍

File: src/Inversion/Impl/Inverter.hs (hunk 13 above)
Comment (jens270996) [reply src/Inversion/Impl/Inverter.hs:25]:
```suggestion
invertFrom (Entry p s)          = Exit p s
```

**Review by Yakokse — APPROVED** [review]
LGTM! :)

**Pull request merged**